File: math-functions/src/main/java/io/github/svbgabriel/FunctionService.java
package io.github.svbgabriel;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.LongStream;

import jakarta.enterprise.context.ApplicationScoped;

@ApplicationScoped
public class FunctionService {

    public ResultFactorialBody factorial(final Long value) {
        if (value == null || value <= 1) {
            return new ResultFactorialBody(BigDecimal.ONE);
        }

        final BigDecimal result = LongStream.rangeClosed(1, value)
                .mapToObj(v -> new BigDecimal(v))
                .reduce(BigDecimal.ONE, (a, b) -> a.multiply(b));

        return new ResultFactorialBody(result);
    }

    public ResultFibonacciBody fibonacci(final Long value) {
        if (value == null) {
            return new ResultFibonacciBody(List.of(BigDecimal.ZERO));
        }

        final List<BigDecimal> results = new ArrayList<>();

        if (value >= 0) {
            results.add(BigDecimal.ZERO);
        }

        if (value >= 2) {
            results.add(BigDecimal.ONE);
        }

        if (value >= 3) {
            BigDecimal n1 = BigDecimal.ZERO;
            BigDecimal n2 = BigDecimal.ONE;
            BigDecimal n3;
            for (int i = 2; i < value; ++i) {
                n3 = n1.add(n2);
                results.add(n3);
                n1 = n2;
                n2 = n3;
            }
        }

        return new ResultFibonacciBody(results);
    }
}

File: math-functions/src/main/kotlin/io/github/svbgabriel/FunctionService.kt
package io.github.svbgabriel

import jakarta.enterprise.context.ApplicationScoped
import java.math.BigDecimal

@ApplicationScoped
class FunctionService {
    fun factorial(value: Long): ResultFactorialBody {
        if (value <= 1) {
            return ResultFactorialBody(BigDecimal.ONE)
        }

        val result = 1L.rangeTo(value)
            .map { it.toBigDecimal() }
            .fold(BigDecimal.ONE) { acc, next -> acc * next  }

        return ResultFactorialBody(result)
    }

    fun fibonacci(value: Long): ResultFibonacciBody {
        val results = mutableListOf<BigDecimal>()

        if (value >= 0) {
            results.add(BigDecimal.ZERO)
        }

        if (value >= 2) {
            results.add(BigDecimal.ONE)
        }

        if (value >= 3) {
            var n1 = BigDecimal.ZERO
            var n2 = BigDecimal.ONE
            var n3: BigDecimal
            for (i in 2 until value) {
                n3 = n1.add(n2)
                results.add(n3)
                n1 = n2
                n2 = n3
            }
        }

        return ResultFibonacciBody(results)
    }
}
